dml/parser_test.go

```diff
@@ -0,0 +1,253 @@
+package dml_test
+
+import (
+	"errors"
+	"testing"
+	"unique"
+
+	"github.com/birdie-ai/golibs/dml"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+)
+
+func TestParser(t *testing.T) {
+	t.Parallel()
+
+	type testcase struct {
+		text string
+		want dml.Stmts
+		err  error
+	}
+
+	for _, tc := range []testcase{
+		{
+			text: "",
+		},
+		{
+			text: "SET",
+			err:  dml.ErrSyntax,
+		},
+		{
+			text: "DELETE",
+			err:  dml.ErrSyntax,
+		},
+		{
+			text: "    SET",
+			err:  dml.ErrSyntax,
+		},
+		{
+			text: "    DELETE",
+			err:  dml.ErrSyntax,
+		},
+		{
+			text: "    SET         ",
+			err:  dml.ErrSyntax,
+		},
+		{
+			text: "    DELETE     ",
+			err:  dml.ErrSyntax,
+		},
+		{
+			text: `SET feedbacks .={} WHERE ["1", "2"];`,
+			err:  dml.ErrSyntax,
+		},
+		{
+			text: `SET feedbacks .=1 WHERE id=1;`,
+			err:  dml.ErrSyntax,
+		},
+		{
+			text: `SET feedbacks .= ["a", "b"] WHERE id=1;`,
+			err:  dml.ErrSyntax,
+		},
+		{
+			text: `SET feedbacks a= [invalid syntax] WHERE id=1;`,
+			err:  dml.ErrSyntax,
+		},
+		{
+			text: `SET feedbacks a.b.c = [invalid syntax] WHERE id=1;`,
+			err:  dml.ErrSyntax,
+		},
+		{
+			text: `SET feedbacks .={} WHERE {invalid JSON};`,
+			err:  dml.ErrSyntax,
+		},
+		{
+			text: `SET feedbacks .={} WHERE {};`, // expect WHERE object with values
+			err:  dml.ErrSyntax,
+		},
+		{
+			text: `SET feedbacks .={} WHERE {"not valid id": 1};`, // expect WHERE object identifier keys
+			err:  dml.ErrSyntax,
+		},
+		{
+			text: `SET feedbacks a=1 WHERE id=[invalid syntax];`,
+			err:  dml.ErrSyntax,
+		},
+		{
+			// expect root assign to have object with ident keys.
+			text: `SET feedbacks .={"not valid key": "test"} WHERE id=1;`,
+			err:  dml.ErrSyntax,
+		},
+		{
+			text: `SET feedbacks .={} WHERE {"id": 1};`,
+			want: dml.Stmts{
+				{
+					Op:     dml.SET,
+					Entity: u("feedbacks"),
+					Assign: dml.Assign{".": map[string]any{}},
+					Where: dml.Where{
+						"id": 1.0,
+					},
+				},
+			},
+		},
+		{
+			text: `SET feedbacks .={} WHERE id=1;`,
+			want: dml.Stmts{
+				{
+					Op:     dml.SET,
+					Entity: u("feedbacks"),
+					Assign: dml.Assign{".": map[string]any{}},
+					Where: dml.Where{
+						"id": 1.0,
+					},
+				},
+			},
+		},
+		{
+			text: `SET feedbacks .={} WHERE id    =   1;`,
+			want: dml.Stmts{
+				{
+					Op:     dml.SET,
+					Entity: u("feedbacks"),
```

Comment:
hmm do you believe the unique thing will make a big difference here ? I only saw it on the release note of some recent Go version...and it reminded me of Lua since it has internalized strings by default/all the time.